java/ql/lib/semmle/code/java/frameworks/JaxWS.qll

```diff
@@ -28,11 +30,106 @@ class JaxWsEndpoint extends Class {
     )
   }
 
-  /** Gets a method annotated with `@WebMethod` or `@WebEndpoint`. */
-  Callable getARemoteMethod() {
+  /**
+   * Gets a method of this class that is not an excluded `@WebMethod`,
+   * and the parameters and return value of which are either of an acceptable type,
+   * or are annotated with `@XmlJavaTypeAdapter`.
+   */
+  Method getARemoteMethod() {
     result = this.getACallable() and
-    exists(AnnotationType a | a = result.getAnAnnotation().getType() |
-      a.hasName(["WebMethod", "WebEndpoint"])
+    not result instanceof InitializerMethod and
+    not exists(Annotation a | a = result.getAnAnnotation() |
+      a.getType().hasQualifiedName(["javax", "jakarta"] + ".jws", "WebMethod") and
+      a.getValue("exclude").(BooleanLiteral).getBooleanValue() = true
+    ) and
+    forex(ParamOrReturn paramOrRet | paramOrRet = result.getAParameter() or paramOrRet = result |
+      exists(Type t | t = paramOrRet.getType() |
+        t instanceof JaxAcceptableType
+        or
+        t.(Annotatable).getAnAnnotation().getType() instanceof XmlJavaTypeAdapter
+        or
+        t instanceof VoidType
+      )
+      or
+      paramOrRet.getInheritedAnnotation().getType() instanceof XmlJavaTypeAdapter
+    )
+  }
+}
+
+/** The annotation type `@XmlJavaTypeAdapter`. */
+class XmlJavaTypeAdapter extends AnnotationType {
+  XmlJavaTypeAdapter() {
+    this.hasQualifiedName(["javax", "jakarta"] + ".xml.bind.annotation.adapters",
+      "XmlJavaTypeAdapter")
+  }
+}
+
+private class ParamOrReturn extends Annotatable {
+  ParamOrReturn() { this instanceof Parameter or this instanceof Method }
+
+  Type getType() {
+    result = this.(Parameter).getType()
+    or
+    result = this.(Method).getReturnType()
+  }
+
+  Annotation getInheritedAnnotation() {
+    result = this.getAnAnnotation()
+    or
+    result = this.(Method).getAnOverride*().getAnAnnotation()
+    or
+    result =
+      this.(Parameter)
+          .getCallable()
+          .(Method)
+          .getAnOverride*()
+          .getParameter(this.(Parameter).getPosition())
```

Comment:
Have you inspected the join-order of this?

Reply:
DCA didn't highlight any bad join order issues, and by evaluating an application that uses `@WebMethod`s there doesn't seem to be any problem either:

```
[2023-08-24 11:22:54] Evaluated non-recursive predicate JaxWS#57655fb8::ParamOrReturn::getInheritedAnnotation#0#dispred#ff@f84667k6 in 152ms (size: 32729).
Evaluated relational algebra for predicate JaxWS#57655fb8::ParamOrReturn::getInheritedAnnotation#0#dispred#ff@f84667k6 with tuple counts:
         24504   ~2%    {2} r1 = JOIN JaxWS#57655fb8::ParamOrReturn#f WITH Annotation#6a559b29::Annotatable::getAnAnnotation#0#dispred#ff ON FIRST 1 OUTPUT Lhs.0, Rhs.1
                    
        230331   ~2%    {2} r2 = JOIN Member#9eba3c33::Method#f WITH JaxWS#57655fb8::ParamOrReturn#f ON FIRST 1 OUTPUT Lhs.0, Lhs.0
                    
        157387   ~0%    {2} r3 = JOIN boundedFastTC:#Member#9eba3c33::Method::overrides#1#dispredPlus#ff:JaxWS#57655fb8::ParamOrReturn#f WITH JaxWS#57655fb8::ParamOrReturn#f ON FIRST 1 OUTPUT Lhs.1, Lhs.0
                    
        387718   ~4%    {2} r4 = r2 UNION r3
         30376   ~2%    {2} r5 = JOIN r4 WITH Annotation#6a559b29::Annotatable::getAnAnnotation#0#dispred#ff ON FIRST 1 OUTPUT Lhs.1, Rhs.1
                    
        171808   ~1%    {3} r6 = SCAN params OUTPUT In.3, In.0, In.2
        171808   ~1%    {3} r7 = JOIN r6 WITH Member#9eba3c33::Callable#f ON FIRST 1 OUTPUT Lhs.0, Lhs.1, Lhs.2
                    
        154808   ~2%    {3} r8 = JOIN r7 WITH Member#9eba3c33::Method#f ON FIRST 1 OUTPUT Lhs.0, Lhs.1, Lhs.2
                    
         87340   ~0%    {3} r9 = JOIN r7 WITH boundedFastTC:#Member#9eba3c33::Method::overrides#1#dispredPlus#ff_10#higher_order_body:_Member#9eba3c33::Callable#f_params#higher_order_body ON FIRST 1 OUTPUT Rhs.1, Lhs.1, Lhs.2
                    
        242148   ~0%    {3} r10 = r8 UNION r9
        412798   ~0%    {3} r11 = JOIN r10 WITH Variable#7a6fdb75::Parameter::getCallable#0#dispred#ff_10#count_range ON FIRST 1 OUTPUT Rhs.1, Lhs.1, Lhs.2
        412798   ~0%    {3} r12 = JOIN r11 WITH JaxWS#57655fb8::ParamOrReturn#f ON FIRST 1 OUTPUT Lhs.0, Lhs.2, Lhs.1
        242148   ~0%    {2} r13 = JOIN r12 WITH params_02#join_rhs ON FIRST 2 OUTPUT Lhs.2, Lhs.0
          2292   ~0%    {2} r14 = JOIN r13 WITH Annotation#6a559b29::Annotatable::getAnAnnotation#0#dispred#ff ON FIRST 1 OUTPUT Lhs.1, Rhs.1
                    
         32668   ~3%    {2} r15 = r5 UNION r14
         57172  ~80%    {2} r16 = r1 UNION r15
                        return r16
```

Do you want me to add any `pragma`s just in case?

Reply:
nah, looks fine.